astropy/units/format/base.py

```diff
@@ -312,3 +303,7 @@ def _decompose_to_known_units(
         raise TypeError(
             f"unit argument must be a 'NamedUnit' or 'CompositeUnit', not {type(unit)}"
         )
+
+    @classmethod
+    def _validate_named_unit(cls, unit: NamedUnit) -> None:
+        cls._validate_unit(unit._get_format_name(cls.name))
```

Comment:
I was hoping that if this method returned the unit instead of `None` then we could simplify `_decompose_to_known_units()`, but turns out that causes test failures.

Reply:
What's the failure?

Reply:
The one failure I would get is

FAILED astropy/units/tests/test_format.py::test_format_styles[cds-erg.Angstrom-1.s-1.cm-2-10000000kg.m-1.s-3] - AssertionError: assert 'erg.AA-1.s-1.cm-2' == 'erg.Angstrom-1.s-1.cm-2'

Basically it boils down to
```python
>>> from astropy import units as u
>>> from astropy.units.format import CDS
>>> u.Angstrom._get_format_name("cds")
'Angstrom'
>>> CDS._units['Angstrom']._get_format_name("cds")
'AA'
```
The last output looks suspicious to me because I don't see "AA" listed neither at https://vizier.unistra.fr/vizier/doc/catstd-3.2.htx nor at https://vizier.unistra.fr/viz-bin/Unit, but figuring out what is going on here shouldn't hold back this refactoring pull request.

Reply:
Agreed.
I wonder if this is possibly a special-character issue  Å?

Reply:
Ah, I missed the very end. And I guess `AA` is the standard ascii-only way to represent that. That also means it is no longer all that obvious which unit is the "preferred" one -- all of them are non-standard (except that `AA` should definitely not be the first one!).`

Probably easiest, though, to just follow `si.py`. I'll make a quick PR.

Reply:
See #17536 - maybe easiest to merge that first so this one can do the further simplification.